test/performance/io/format_sam_benchmark.cpp

```diff
@@ -0,0 +1,225 @@
+            auto align_result = *(seqan3::align_pairwise(std::tie(query, reference), config).begin());
+            std::string const current_query_id = default_query_id + std::to_string(i);
+
+            sam_out.emplace_back(query,                                  // field::seq
+                                 current_query_id,                       // field::id
+                                 align_result.front_coordinate().first,  // field::offset
+                                 reference,                              // field::ref_seq
+                                 reference_id,                           // field::ref_id
+                                 align_result.front_coordinate().second, // field::ref_offset
+                                 align_result.alignment(),               // field::alignment
+                                 align_result.score(),                   // field::mapq
+                                 qualities,                              // field::qual
+                                 seqan3::sam_flag::none);                // field::flag
+        }
+
+        file_dict[n_queries] = stream.str();
+    }
+
+    return file_dict[n_queries];
+}
+
+void write_file(std::string file_name, size_t n_queries)
+{
+    std::ofstream ostream{file_name};
+    ostream << create_sam_file_string(n_queries);
+    ostream.close();
+}
+
+// ============================================================================
+// seqan3
+// ============================================================================
+
+void sam_file_read_from_stream(benchmark::State &state)
+{
+    size_t n_queries = state.range(0);
+
+    std::istringstream istream{create_sam_file_string(n_queries)};
+
+    for (auto _ : state)
+    {
+        // refill stream (same constant for seqan2 benchmark)
+        istream.clear();
+        istream.seekg(0, std::ios::beg);
+
+        seqan3::alignment_file_input fin{istream, seqan3::format_sam{}};
+
+        // read all records and store in internal buffer
+        auto it = fin.begin();
+        while (it != fin.end())
+            ++it;
+    }
+}
+
+void sam_file_read_from_disk(benchmark::State &state)
+{
+    size_t n_queries = state.range(0);
+    seqan3::test::tmp_filename file_name{"tmp.sam"};
+    auto tmp_path = file_name.get_path();
+
+    write_file(file_name.get_path(), n_queries);
+
+    for (auto _ : state)
+    {
+        seqan3::alignment_file_input fin{tmp_path};
+
+        // read all records and store in internal buffer
+        auto it = fin.begin();
+        while (it != fin.end())
+            ++it;
+    }
+}
+
+#if SEQAN3_HAS_SEQAN2
+// ============================================================================
+// seqan2 read from stream
+// ============================================================================
+
+void seqan2_sam_file_read_from_stream(benchmark::State &state)
+{
+    size_t n_queries = state.range(0);
+    seqan3::test::tmp_filename file_name{"tmp.sam"};
```

Comment:
I am not too familiar with the seqan2/seqan3 mixed support, is it okay to have a seqan3 functionality in a seqan2 function?

Reply:
Yes it is. You can use seqan3 and seqan side by side (if they don't cause trouble). And in this case it's not part of the performance measurement, so I am also not cheating :p